packages/celotool/src/cmds/bots/auto-vote.ts

```diff
@@ -0,0 +1,204 @@
+      const lockedGold = await kit.contracts.getLockedGold()
+
+      console.info('Determining which groups have capacity for more votes')
+      const groupCapacities = new Map<string, BigNumber>()
+      for (const groupAddress of await validators.getRegisteredValidatorGroupsAddresses()) {
+        const vgv = await election.getValidatorGroupVotes(groupAddress)
+        if (vgv.eligible) {
+          groupCapacities.set(groupAddress, vgv.capacity)
+        }
+      }
+
+      console.info('Calculating weights of groups based on the scores of elected validators')
+      const groupWeights = await calculateGroupWeights(kit)
+
+      for (const key of botKeysVotingThisRound) {
+        const botAccount = ensure0x(getAccountAddressFromPrivateKey(key))
+
+        kit.addAccount(key)
+        kit.defaultAccount = botAccount
+
+        console.info(`Voting as: ${botAccount}.`)
+        try {
+          const lockedGoldAmount = await lockedGold.getAccountTotalLockedGold(botAccount)
+          if (lockedGoldAmount.isZero()) {
+            console.info(`No locked gold exists for ${botAccount}, skipping...`)
+            continue
+          }
+
+          const currentVotes = (await election.getVoter(botAccount)).votes
+          const currentGroups = currentVotes.map((v) => v.group)
+
+          const randomlySelectedGroup = getWeightedRandomChoice(
+            groupWeights,
+            [...groupCapacities.keys()].filter((k) => {
+              const capacity = groupCapacities.get(k)
+              return capacity && capacity.isGreaterThan(0) && !currentGroups.includes(k)
+            })
+          )
+
+          // Revoke existing vote(s) if any
+          for (const vote of currentVotes) {
+            const revokeTxs = await election.revoke(
+              botAccount,
+              vote.group,
+              vote.pending.plus(vote.active)
+            )
+            for (const tx of revokeTxs) {
+              await tx.sendAndWaitForReceipt({ from: botAccount })
+            }
+          }
+
+          const groupCapacity = groupCapacities.get(randomlySelectedGroup)!
+          const voteAmount = BigNumber.minimum(lockedGoldAmount, groupCapacity)
+          const voteTx = await election.vote(randomlySelectedGroup, BigNumber.minimum(voteAmount))
+          await voteTx.sendAndWaitForReceipt({ from: botAccount })
+          console.info(`Completed voting as ${botAccount}`)
+
+          groupCapacities.set(randomlySelectedGroup, groupCapacity.minus(voteAmount))
+        } catch (error) {
+          console.error(`Failed to vote as ${botAccount}`)
+          console.info(error)
+        }
+      }
+    }
+  } catch (error) {
+    console.error(error)
+  } finally {
+    process.exit(0)
+  }
+}
+
+async function calculateGroupWeights(kit: ContractKit): Promise<Map<string, BigNumber>> {
+  const election = await kit.contracts.getElection()
+  const validators = await kit.contracts.getValidators()
+
+  const groupScores = new Map<string, BigNumber[]>()
```

Comment:
What do you think about using `lodash`'s `groupBy` and `mapValues` for a more concise version of this? 

```typescript
  const validatorSigners = await election.getCurrentValidatorSigners()
  const validatorAccounts = await concurrentMap(
    10,
    validatorSigners,
    validators.getValidatorFromSigner
  )

  const validatorsByGroup = groupBy(validatorAccounts, (validator) => validator.affiliation!)
  const validatorGroupScores = mapValues(validatorsByGroup, (vals) =>
    vals.reduce((a, b) => a.plus(b.score), new BigNumber(0))
  )

  return new Map(Object.entries(validatorGroupScores))

```

Reply:
That looks much better. Thanks!

Reply:
hmm... actually this doesn't totally solve it since what needs to happen in calculating the group score is to find the _average_ score of all the validators in the group. Your suggestion gets the sum of them all. Perhaps there's still a use for `groupBy` and `mapValues` though.